src/events/kafka.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Supporting objects for Karapace-Kafka relation."""
+
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.data_interfaces import (
+    BootstrapServerChangedEvent,
+    KafkaRequirerEventHandlers,
+    TopicCreatedEvent,
+)
+from ops import Object, RelationBrokenEvent
+
+from literals import KAFKA_REL, Status
+
+if TYPE_CHECKING:
+    from charm import KarapaceCharm
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaHandler(Object):
+    """Implements the requirer-side logic for client applications relating to Kafka."""
+
+    def __init__(self, charm) -> None:
+        super().__init__(charm, "kafka_client")
+        self.charm: "KarapaceCharm" = charm
+
+        self.kafka = KafkaRequirerEventHandlers(
+            self.charm, relation_data=self.charm.context.kafka_requirer_interface
+        )
+
+        self.framework.observe(self.charm.on[KAFKA_REL].relation_broken, self._on_kafka_broken)
+        self.framework.observe(
+            getattr(self.kafka.on, "bootstrap_server_changed"),
+            self._on_kafka_bootstrap_server_changed,
+        )
+        self.framework.observe(
+            getattr(self.kafka.on, "topic_created"), self._on_kafka_topic_created
+        )
+
+    def _on_kafka_bootstrap_server_changed(self, event: BootstrapServerChangedEvent) -> None:
+        """Handle the bootstrap server changed."""
+        # Event triggered when a bootstrap server was changed for this application
+        logger.info(f"Bootstrap servers changed into: {event.bootstrap_server}")
+        self.charm.on.config_changed.emit()
+
+    def _on_kafka_topic_created(self, _: TopicCreatedEvent) -> None:
+        """Handle the topic created event."""
+        self.charm.config_manager.generate_config()
+        self.charm.workload.start(layer=self.charm.karapace_layer)
```

Comment:
**todo** Can we move the definition of the karapace layer into the workload, such that when we can just call `workload.start()`?

Otherwise here the two codes (VM and K8s) starts to diverge

Reply:
I'm reluctant to move layer away from `charm.py`. We've needed to pass conditional parameters before, that might depend on relations/databags. It's not a problem at the moment for Karapace, but I would personally keep it in charm.

Reply:
How would you make the `src/events` and `src/managers` consistent between VM and K8s then?

Reply:
Yes, this is an inconsistency. But if we need to make a parameter conditional based on information from `cluster/models` we also land on a problem where layer is not a static object that can be just declared on workload.

Reply:
I just checked the spark charms, and we indeed keep the layers definition in the workload, see [here](https://github.com/canonical/spark-history-server-k8s-operator/blob/main/src/workload.py#L47). 

I would rather avoid the inconsistency and diverge VM and K8s (if needed) once that we have the actual use-case at hand. I don't see the point of diverging already right now

but let's see what others think

Reply:
Moved it to workload. We can move it around if needed in the future.